src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -93,61 +93,250 @@ let check_uvars r t =
+          move_decreases d (pfx@sfx) (pfx'@sfx')
+        | Some (pfx, DECREASES d, sfx), None ->
+          move_decreases d (pfx@sfx) (U.comp_flags c')
+        | _ -> failwith "Impossible"
+  in
+  let extract_annot_from_body (lbtyp_opt:option<typ>)
+    : typ
+    * term
+    * bool
+    = let rec aux e
+        : typ * term * bool
+        = let e = SS.compress e in
+          match e.n with
+          | Tm_meta(e', m) ->
+            let t, e', recheck = aux e' in
+            t, { e with n = Tm_meta(e', m) }, recheck
+
+          | Tm_ascribed(e', (Inr c, tac_opt), lopt) ->
+            if U.is_total_comp c
+            then let t, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type (U.comp_result c) lbtyp_opt in
+                 let e = { e with n = Tm_ascribed(e', (Inr (S.mk_Total t), tac_opt), lopt) } in
+                 lbtyp, e, recheck
+            else raise_error (Errors.Fatal_UnexpectedComputationTypeForLetRec,
+                              BU.format1 "Expected a 'let rec' to be annotated with a value type; got a computation type %s"
+                                                   (Print.comp_to_string c))
+                              rng
+
+          | Tm_ascribed(e', (Inl t, tac_opt), lopt) ->
+
+            let t, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type t lbtyp_opt in
+            let e = { e with n = Tm_ascribed(e', (Inl t, tac_opt), lopt) } in
+            lbtyp, e, recheck
+
+          | Tm_abs _ ->
+            let bs, body, rcopt = U.abs_formals_maybe_unascribe_body false e in
+            let mk_comp t =
+              if Options.ml_ish()
+              then U.ml_comp t t.pos
+              else S.mk_Total t
+            in
+            let mk_arrow c = U.arrow bs c in
+            let rec aux body =
+              let body = SS.compress body in
+              match body.n with
+              | Tm_meta (body, m) ->
+                let t, body', recheck = aux body in
+                let body = { body with n = Tm_meta(body', m) } in
+                t, body, recheck
+
+              | Tm_ascribed (_, (Inl t, _), _) -> //no decreases clause here
+                begin
+                match lbtyp_opt with
+                | Some lbtyp ->
+                  lbtyp, body, false
+
+                | None ->
+                  let t = mk_arrow (mk_comp t) in
+                  t, body, true
+                end
+
+              | Tm_ascribed (body', (Inr c, tac_opt), lopt) ->
+                let tarr = mk_arrow c in
+                let tarr, lbtyp, recheck = reconcile_let_rec_ascription_and_body_type tarr lbtyp_opt in
+                let bs', c = un_arrow tarr in
+                if List.length bs' <> List.length bs
+                then failwith "Impossible"
+                else let subst = U.rename_binders bs' bs in
+                     let c = SS.subst_comp subst c in
+                     let body = { body with n = Tm_ascribed(body', (Inr c, tac_opt), lopt) } in
```

Comment:
This could potentially add a redundant ascription, right? E.g. when `lbtyp` is not set, we will add `lbtyp` (around line 216) and also keep the ascription here, which means a redundant subtyping query. I wonder if we could remove the ascription in this case.

Reply:
I don't want to remove this ascription, for two reasons:

1. I don't want to disturb the term that the user wrote. If they wrote an ascription there (perhaps with a very careful tac_opt too) then we should keep it.

2. The ascribed type may not be identical to the lbtyp. For example, see the negative test case in Dec.fst where the ascribed type has a stronger precondition than the val. Keeping the ascription here allows us to detect that case.

Reply:
I was thinking of the case where there is no `val`, and there is a `let f (x:a) : Tot b (decreases e)` .. in this case, at the end of this function we would have an ascribed let binding and also lbtyp set to the same type. So we would do a redundant subtyping query for the same type?

Reply:
(Of course when both the ascription and lbtyp are present to begin with, we should retain both.)

Reply:
We spoke offline: The goal here was to not remove any checks that are present in the existing code. So, we'll keep the check. Given discussion below about not re-typechecking lbtyp in some cases, trying to optimize and remove a check here is particularly slippery.